TIP-24 Tangle Block (Stardust) (#55)
iotaledger/tips: tips/TIP-0024/tip-0024.md

# Motivation

To better understand this layered design, consider the Internet Protocol (IP), for example: There is an Ethernet frame that contains an IP payload. This in turn contains a TCP packet that encapsulates an HTTP payload. Each layer has a certain responsibility and once this responsibility is completed, we move on to the next layer.

The same is true with how blocks are parsed. The outer layer of the block enables the mapping of the block to a vertex in the Tangle and allows us to perform some basic validation. The next layer may be a transaction that mutates the ledger state, and one layer further may provide some extra functionality on the transactions to be used by applications.

By making it possible to add and exchange payloads, an architecture is being created that can easily be extended to accommodate future needs.

# Specification

## Block ID

The *Block ID* is the [BLAKE2b-256](https://tools.ietf.org/html/rfc7693) hash of the entire serialized block.

## Serialized Layout

> The following table describes the serialization of a _Block_ following the notation from [TIP-21](../TIP-0021/tip-0021.md):

Reviewer: Broken link

Reviewer 2: It will work once merged.

      </details>
    </td>
  </tr>
  <tr>
    <td>Payload Length</td>
    <td>uint32</td>
    <td>The length of the following payload in bytes. A length of 0 means no payload will be attached.</td>
  </tr>
  <tr>
    <td valign="top">Payload <code>optOneOf</code></td>
    <td colspan="2">
      <details open="true">
        <summary>Generic Payload</summary>
        <blockquote>
          An outline of a generic payload
        </blockquote>
        <table>
          <tr>
            <th>Name</th>
            <th>Type</th>
            <th>Description</th>
          </tr>
          <tr>
            <td>Payload Type</td>
            <td>uint32</td>
            <td>
              The type of the payload. It will instruct the node how to parse the fields that follow.
            </td>
          </tr>
          <tr>
            <td>Data Fields</td>
            <td>ANY</td>
            <td>A sequence of fields, where the structure depends on <code>Payload Type</code>.</td>
          </tr>
        </table>
      </details>
  <tr>
    <td>Nonce</td>
    <td>uint64</td>
    <td>The nonce which lets this block fulfill the PoW requirement.</td>
  </tr>
</table>

## Syntactic validation

> The Tangle can only contain syntactically valid blocks. Invalid must be rejected by the node.

Reviewer: ```suggestion
The Tangle can only contain syntactically valid blocks. Invalid blocks must be rejected by the node.
```

- There must be no trailing bytes after all block fields have been parsed.
  
### PoW validation

The PoW that needs to be performed for each block protects the network against denial-of-service attacks where in a short time too many blocks are issued for the nodes to process. As the processing time of a block heavily depends on the contained payload, the PoW check can also depend on the `Payload Type` and is described under [Payloads](#payloads).
It is important to note, that the actual parsing and validating of a payload can be computationally expensive. Thus, it is recommended to first parse the block with all its fields including `Payload Type` (but not parsing or validating the actual payload `Data Fields`). Now, simple syntactic validation steps &ndash; including PoW validation &ndash; can be performed and invalid blocks already filtered out before the payload is validated. With this approach, payload-based PoW validation is not significantly more expensive than payload-agnostic validation.

## Payloads

> While blocks without a payload, i.e. `Payload Length` set to zero, are valid, such blocks do not contain any information. As such, blocks usually contain a payload. The detailed specification of each payload type is out of scope of this TIP. The following table lists all currently specified payloads that can be part of a block and links to their specification:

Reviewer: `all` is not completely correct as treasury transaction is currently still part of stardust

Reviewer 2: A treasury transaction may only be put in a milestone receipt if I'm not mistaken.

Reviewer: Yeah ignore me, I'm tired and already a bit on holidays 🤦🏻

| Payload Name | Type Value | TIP                               | PoW Condition                  |
| ------------ | ---------- | --------------------------------- | ------------------------------ |
| No Payload   | -          | -                                 | PoW score ≥ `Min PoW Score`    |
| Tagged Data  | 5          | [TIP-23](../TIP-0023/tip-0023.md) | PoW score ≥ `Min PoW Score`    |
| Transaction  | 6          | [TIP-20](../TIP-0020/tip-0020.md) | PoW score ≥ `Min PoW Score`    |
| Milestone    | 7          | [TIP-29](../TIP-0029/tip-0029.md) | `nonce` = `0x0000000000000000` |

## Example

Below is the full serialization of a valid block with a _Tagged Data Payload_. The tag is the "IOTA" ASCII string and the data is the "hello world" ASCII string. Bytes are expressed as hexadecimal numbers.

> - Protocol Version (1-byte): `00` (0)

Reviewer: Should probably put protocol version 2, I don't think this block would have been valid with version 0

---
tip: 24
title: Tangle Message
description: Generalization of the Tangle transaction concept
author: Wolfgang Welz (@Wollac) <wolfgang.welz@iota.org>
discussions-to: https://github.com/iotaledger/tips/pull/55
status: Draft
type: Standards

> layer: Core

Reviewer 2: ```suggestion
layer: Core
replaces: 6
```

          </tr>
        </table>
      </details>
  <tr>
    <td>Nonce</td>
    <td>uint64</td>
    <td>The nonce which lets this message fulfill the PoW requirement.</td>
  </tr>
</table>

## Syntactical validation

The following criteria defines whether the message passes the syntactical validation:

- The total length of the serialized message must not exceed `Max Message Length`.
- `Network ID` must match the first 8 bytes of the BLAKE2b-256 hash of the `Network String` parameter.
- Parents:
  - `Parents Count` must be at least 1 and not larger than `Max Parents Count`.
  - `Parents` must be sorted in lexicographical order.
  - Each `Message ID` must be unique.
- Payload (if present):
  - `Payload Type` must match one of the values described under [Payloads](#payloads).
  - `Data fields` must be correctly parsable in the context of the `Payload Type`.
  - The payload itself must pass syntactic validation.
- `Nonce` must be a valid solution of the message PoW as described in [TIP-12](../TIP-0012/tip-0012.md). The resulting PoW score must be larger or equal `Min PoW Score`.
- There must be no trailing bytes after all message fields have been parsed.

## Payloads

> While messages without a payload, i.e. `Payload Length` set to zero, are valid, such messages do not contain any information. As such, messages usually contain a payload. The detailed specification of each payload type is out of scope of this RFC. The following table lists all currently specified payloads that can be part of a message and links to their specification:

Reviewer 2: ```suggestion
While messages without a payload, i.e. `Payload Length` set to zero, are valid, such messages do not contain any information. As such, messages usually contain a payload. The detailed specification of each payload type is out of scope of this TIP. The following table lists all currently specified payloads that can be part of a message and links to their specification:
```

| ------------ | ---------- | --------------------------------------- |
| Transaction  | 0          | [TIP-7](../TIP-0007/tip-0007.md)        |
| Milestone    | 1          | [TIP-8](../TIP-0008/tip-0008.md)        |
| Tagged Data  | 5          | [draft TIP-23](../TIP-0023/tip-0023.md) |

### Tagged data payload

This payload allows the addition of arbitrary data to the encapsulationg messages. An optional tag can be used to categorize the data. It is important to note that the tag is not considered by the protocol, it serves just as a marker for second layer applications.

| Name         | Type                  | Description                                          |
| ------------ | --------------------- | ---------------------------------------------------- |
| Payload Type | uint32                | Set to *value 5* to denote an _Tagged Data Payload_. |
| Tag Length   | uint8                 | The length of the following tag field in bytes.      |
| Tag          | ByteArray[Tag Length] | The tag of the data.                                 |
| Data         | ByteArray             | Binary data.                                         |

> *Syntactic validation*: `Tag Length` must not be larger than `Max Tag Length`. When `Tag Length` is 0, the `Tag` has length 0 and is omitted.

Reviewer 2: Since you reference this payload in the table above, I don't think that this section is needed here.

Author: copy-paste mistake :see_no_evil:

## Example

> Below is the full serialization of a valid message with an indexation payload. The index is the "IOTA" ASCII string and the data is the "hello world" ASCII string. Bytes are expressed as hexadecimal numbers.

Reviewer 2: ```suggestion
Below is the full serialization of a valid message with a tagged data payload. The tag is the "IOTA" ASCII string and the data is the "hello world" ASCII string. Bytes are expressed as hexadecimal numbers.
```
